### PR TITLE
Add ASYNC support

### PR DESCRIPTION
**Pull request by mattcaswell**
Add support for performing crypto operations asynchronously - as per my post to openssl-dev@openssl.org.

### CODE REVIEW

**Comment by mattcaswell** [comment]
I just pushed a new version to respond to feedback received so far (also rebased against latest master).

File: crypto/cryptlib.c [hunk 1]
```diff
@@ -231,16 +231,63 @@ BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD fdwReason, LPVOID lpvReserved)
...
+
+    return data[idx];
+}
+
+int CRYPTO_set_thread_local(int idx, void *locdata)
+{
+    void **data;
+
+    if (idx < 0 || idx >= CRYPTO_THREAD_LOCAL_TOTAL_VALS)
+        return NULL;
```
Comment (stevelinsell) [inline crypto/cryptlib.c]: Should this not explicitly return 0 on error rather than NULL? Copy and paste error?

Comment (mattcaswell) [inline crypto/cryptlib.c]: You mean in CRYPTO_set_thread_local? Oops. Yes. Good catch.

**Comment by mattcaswell** [comment]
Fixed a few issues. Travis is now showing green across all builds :-)

File: engines/e_dasync.c [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+{
+    dummy_pause_job();
+    return RSA_PKCS1_SSLeay()->rsa_priv_enc(flen, from, to, rsa, padding);
+}
+
+static int dasync_rsa_priv_dec(int flen, const unsigned char *from,
+                      unsigned char *to, RSA *rsa, int padding)
+{
+    /* Ignore errors - we carry on anyway */
+    ASYNC_pause_job();
```
Comment (stevelinsell) [inline engines/e_dasync.c]: Is there a reason why this particular function has not been moved to use dummy_pause_job() like the others in dasync engine? Sorry for flagging this late, just spotted it while looking at something else!

**Comment by stevelinsell** [comment]
I was trying out this code and observed a problem I wanted to feedback. I tried calling an ECDSA_sign operation using the dasync engine and running 2 simultaneous jobs. The dasync engine obviously does not have async enabled for ecdsa (so I expected it to run 1 job until ASYNC_FINISHED and then run the second job within my loop) but interestingly it *is* pausing jobs because a hash is done as part of generating some random numbers within ECDSA_sign and the sha1 hash is async enabled in the engine. This in itself works fine but as part of the random number generation the CRYPTO_LOCK_RAND is being got but is not released when the job pauses. This then causes a deadlock when the second job tries to get the CRYPTO_LOCK_RAND. I'm not sure what to suggest as the best way to deal with this scenario.

File: engines/e_dasync.c (hunk 2 above)
Comment (mattcaswell) [inline engines/e_dasync.c]: No, it just got missed. I've resolved this in my latest version.

**Comment by mattcaswell** [comment]
Added some commits to deal with the rand deadlock problem. @stevelinsell can you check that this resolves the issue in your test code?

**Comment by mattcaswell** [comment]
Updated to deal with a number of thread-local-storage related issues. In particular this now uses pthreads rather than the less portable "__thread" syntax for the POSIX implementation. Additionally there have been a number of changes to ensure that the thread-local stuff works for statically linked windows.
This has had some API impacts around initialisation and cleanup. ASYNC_init_pool and ASYNC_free_pool have gone. They are replaced by ASYNC_init, ASYNC_init_thread, ASYNC_cleanup_thread and ASYNC_cleanup. Documentation has been updated accordingly.

**Comment by richsalz** [comment]
Are you planning on eventually moving everything over to use standard thread-local storage?

**Comment by mattcaswell** [comment]
Not quite sure what you mean by "standard thread-local storage". Currently there are two types of thread-local storage in use: pthreads and windows - both are needed. If pull request #451 eventually gets implemented then I will move stuff over to use that instead.

**Comment by richsalz** [comment]
We can get rid of various hash tables that index thread-id -> some state; e.g., ERR_STATE.  Basically all uses of CRYPTO_THREADID and friends.

**Comment by mattcaswell** [comment]
Right...that's what #451 is about. If that pull request goes ahead then yes, the async stuff will be updated to use it.

**Comment by richsalz** [comment]
That’s what I wanted to hear, thanks.  In my view 451 isn’t ready yet; it does some things wrong.  Maybe I’ll talk with the author.

**Comment by mattcaswell** [comment]
This has now been merged. Closing.

**Pull request closed without merging**